docs/config.md
# Configuration

## Default configuration file

```toml
# Default config file:
# Path to subscribed tinylogs:
subscribed_data = "~/.config/gtl/subs"
# Refresh time:
refresh = 10
# Date display format
date_format = "Mon 02 Jan 2006 15:04 MST"
# Log file:
log_file = "/dev/null"
# Optional: Highlight when text is found in content.
# Separate values by a coma, eg:
# highlights = "@bacardi55, @bacardi, anything"
# Ignore Highlights from author:
# Must be the full author displayed name!
# ignore_author_highlights = "🤔 @bacardi55"
# Maximum number of entries showed in cli mode. If --limit is used, it will overide this setting.
# Will be ignored in tui mode.
cli_limit = 10
# Mode: either cli, tui or gemini
mode = "tui"
# If false, standard ascii characters will be used.
tui_status_emoji = false

# Enable tinylog edition from gtl:
# This will use an external editor,
# configured in your EDITOR environment variable.
# You can check with 'echo $EDITOR' to see if it
# is configured correctly.
# ctrl+n is disabled when set to false.
# Settings available since v0.5.0
allow_edit = false
# Path to tinylog file. This option is ignored if
# allow_edit = false.
# If not a valid file, editing will not be possible
# and ctrl+n will be disabled.
tinylog_path = "path/to/tinylog/file.gmi"
# Path to script to be executed after the edition is done.
# This script needs to be executable.
# If not a valid executable script, it will be ignored.
post_edit_script = "path/to/script"
# Auto refresh feeds after editing the tinylog file.
# Only used when allow_edit = true
# Settings available since v0.5.1
post_edit_refresh = false
# Limit the number of entries displayed in TUI.
# Indicate 0 for all entries.
# Settings available since v0.6.0
tui_max_entries = 0
# Copy a pre formatted text to clipboard when creating a new entry
# On linux, requires 'xclip' or 'xsel'
# Settings available since v0.6.0
tui_copy_stub_clipboard = false
# If you are running gtl without X, the copy to clipboard feature
# will not work (Or if you don't have `xclip` or `xsel`).
# In this case, enabling this option will allow gtl to display
# the sub text in a modal for easy copy in tools like tmux
# At this stage, the rendering is ugly because of an issue in cview:
# https://code.rocketnine.space/tslocum/cview/issues/72#issuecomment-3968
tui_show_stub = false
# TUI Theming:
# Please see the TUI and theming documentation page in docs/mode-tui.md and docs/theming.md
# Dracula theme example:
#tui_color_background = "282a36"
#tui_color_text = "f8f8f2"
#tui_color_author_name = "ffb86c"
#tui_color_links = "8be9fd"
#tui_color_box = "44475a"
#tui_color_focus_box = "f8f8f2"
#tui_color_elapsed_time = "f1fa8c"
#tui_color_selected_background = "f8f8f2"
#tui_color_selected_foreground = "282a36"
#tui_color_highlight = "ff79c6"
#tui_color_quote = "f1fa8c"
#tui_color_button_color = "44475a"
#tui_color_button_text = "f8f8f2"
#tui_color_button_focus = "ffb86c"
#tui_color_button_focus_text = "282a36"

# Bookmarks (only available via TUI for now):
# Enable/disable bookmarks feature:
bookmarks_enabled = false
# Path to bookmarks file:
bookmarks_file = "~/.config/gtl/bookmarks.json"
# Reverse order (from newer to older) instead of default (older to newer):
bookmarks_reverse_order = true
```

By default, gtl will look for ~/.config/gtl/gtl.toml . It will create it if needed.

The --config option only look for the file, it will not create it if the file given as argument of --config doesn't exist.

## Subscription management

You can add and remove tinylog entry either manually from the file directly, or use gtl to do it for you:
```
Subscription management usage:
	add --url url [--title title]	Indicate a new tinylog url (and optional title) to subscribe to.
	rm --url url			Indicate a tinylog url to be removed from the subscription.
```



CHANGELOG.md
CHANGELOG

## v1.0.0:

**Major Release**

### New features:
* Implements #34: Add bookmarks management
* Implements #51: Ignore highlight list

### Bug Fixes:
* Fix #50: World between bracket are not displayed
* Fix #52 part1: highlight bar in the subscription list does not change color
* Fix #52 part2: color of the buttons in the pop-up window does not change color
* Fix #52 part3: Wrong links in link modal when multiple links
* Fix #42: Entries with ## within a preformated text are failing

## v0.7.0:

### New features:
* Implements #44: Custom theme. See the documentation in [docs/mode-tui-theming.md]
* Implements #47: Search. Shortcut `/` will open a modal asking for a search. It will basically filter entries that contains exactly your search (case insensitive). It isn’t fuzzy or advance searching (more filtering)
* Implements #48: Add button to go to author tinylog on action modal (alt+enter)
* Implements #40: Add tinylog url in the gemini mode output (used for generating gemini://tinylogs.gmi.bacardi55.io)

### Bug Fixes
* Fix 42: Hide new line separator in response stub
* Fix subscription name with space in their name
* Fix #45: Crash when empty line in subscription file
* Fix #49: Allow link without space between `=>` and the actual link.

## v0.6.0:

**Major release**

### Breaking changes:
* Shortcut to create a new entry is now "N" (uppercase) and no longer ctrl+n.

### New features:
* Implements region in timeline TextView: This will allow to navigate between entries using J and K (uppercase)
* #28: stubs for new entry: Clicking on "N" (uppercase) will copy in the clipboard a stub for a new entry with the current date if the option `tui_copy_stub_clipboard` is set (see README).
* #28: stubs for response: When selecting entry with J/K, using "R" (uppercase) will copy a pre-formatted response in your clipboard if the option `tui_copy_stub_clipboard` is set (see README).
* #33: Open links in entries, relates to #23. When using J/K to select an entry, using "O" will automatically open the link in the entry. If more than one link is in the entry, a popup will ask for one (or all) link(s) to open.
* #37: Implement a gemini compatible output
* #36: Show the tinylog entry being replied to: When selecting an entry with J/K, if the entry is a response to another tinylog entry and in the RFC format for a response, it will show the original entry in a popup when using "T" (uppercase)?
* #35: parser v2: Improve parser to be compatible with proposition 15 of the RFC (https://codeberg.org/bacardi55/gemini-tinylog-rfc/issues/15) that allow multiple break line in posts. Also improve parsing in general.
* Implement a modal for entry details via alt+enter
* Implement `tui_show_stub` option to show the stub to copy instead (or additionally) of copying it to the clipboard. Can be useful if you run gtl without X and/or are using tools like tmux to copy texts. Idea is from @szczezuja.
* Implement entry selection via mouse left click and simplify highlights code.

### Smaller improvements:
* [TUI] Add optional limit to number of displayed entries: Allow to limit the number of entries displayed in the timeline. See `tui_max_entries` option in README.
* Use gemini://tinylogs.gmi.bacardi55.io/ as default subscripton when no subscription file is found.
* Update dependencies
* Remove dead code

### Bug fixes
* Improve gemtext formating for level 3 headers and lists.
* Fix #29: Timeline refresh was previous refresh
* Fix error for missing file: if a tinylog isn't available on a working capsule, status was wrong.
* Fix limit bug in cli mode
* Remove date format duplicate

### Known Issues:
* Text in modal is not nicely formatted (eg: Thread or reply stub). This is due to an issue in cview: https://code.rocketnine.space/tslocum/cview/issues/72

Please read the README that has been updated accordingly.

## 0.5.2

**Critical Bug Fixing release**

* Fix case sensitivity in avatar/author metadata
* Fix date parsing error

## 0.5.1

*Minor release*

* Improve main panel title by adding filtered tinylog name
* Remove modal message after editing tinylog if no post script is defined.
* Remove message post script if successful and add config to refresh feeds after editing entry
* Improve focus changes
* Improve gemtext formatting for responses.
* Harmonized terminology around tinylogs and improve panels title

## 0.5.0

**Major release**

* Add ability to edit tinylog and launch a post edit script

## 0.4.8

*Minor release*

* Add UTC offset for date format
* Resolves #21: [TUI] Add a refresh message while feeds are refreshed
* Resolves #16: [TUI] Add ability to filter out tinylogs

## 0.4.7

*Minor release*

**Breaking change**: emoji status is now an option and disabled by default.
To enable it tui_status_emoji = true in gtl.toml

* Add sidebar toggle command in help box
* Improve title readability
* Add option for emoji usage. Also fixed some gemtext formatting
* fix gemtext formatting

## 0.4.6

*Minor release*

* Reformat TUI Sub list, resolves #15
* Resolves #17: [TUI] Don't quit gtl when displaying help with q
* change dead link emoji
* Add shortcut 's' to toggle sidebar display
* Add formating for links, lists and quotes
* Update sidebar color to match timeline color formatting
* Add help message in All sub secondary text
* Add new date format…

## 0.4.5

*Minor release*

**Breaking changes:**
* TUI shortcuts have been changed to single letter, no needs to ctrl anymore. Look at the README or display the help in TUI mode with ?.

* Fix #1: Create config dir at first launch and fix first launch
* Default mode is now TUI
* TUI - Implements #9: Show tinylogs with 0 compatible entries
* TUI - Resolves #12: Add last refresh in timeline. Hide footer for now before it is used again
* TUI - Resolves #13: Add color on border for focus panel
* TUI - Add display/hide help box and remove header
* TUI - shortcuts have changed, press ? for the help

## 0.4.3

*Minor release*

* fix focus
* Add TAB in usage

config/config.go
package config

import (
	"fmt"
	"log"
	"os"
	"path/filepath"

	// TODO: Remove ioutil dependencies since go ≥ 1.16.
	"io/ioutil"

	"github.com/mitchellh/go-homedir"
	"github.com/pelletier/go-toml"

	"git.bacardi55.io/bacardi55/gtl/core"
)

var defaultConf = []byte(`# Default config file
## General configuration:
# Path to subscribed tinylogs:
subscribed_data = "~/.config/gtl/subs"
# Date display format
date_format = "Mon 02 Jan 2006 15:04 MST"
# Log file:
log_file = "/dev/null"
# Optional: Highlight when text is found in content.
# Separate values by a coma, eg:
# highlights = "@bacardi55, @bacardi, anything"
# Ignore Highlights from author:
# Must be the full author displayed name!
# ignore_author_highlights = "🤔 @bacardi55"
# Maximum number of entries showed in cli mode. If --limit is used, it will overide this setting.
# Will be ignored in tui mode.
cli_limit = 10
# Mode: either cli or tui
mode = "tui"
# If false, standard ascii characters will be used.
tui_status_emoji = false

## TUI options:
# Enable tinylog edition from gtl:
# This will use an external editor,
# configured in your EDITOR environment variable.
# You can check with 'echo $EDITOR' to see if it
# is configured correctly.
# ctrl+n is disabled when set to false.
allow_edit = false
# Path to tinylog file. This option is ignored if
# allow_edit = false.
# If not a valid file, editing will not be possible
# and ctrl+n will be disabled.
tinylog_path = "path/to/tinylog/file.gmi"
# Path to script to be executed after the edition is done.
# This script needs to be executable.
# If not a valid executable script, it will be ignored.
post_edit_script = "path/to/script"
# Auto refresh feeds after editing the tinylog file.
# Only used when allow_edit = true
post_edit_refresh = false
# Limit the number of entries displayed in TUI.
# Indicate 0 for all entries.
tui_max_entries = 0
# Copy a pre formatted text to clipboard when creating a new entry
# On linux, requires 'xclip' or 'xsel'
tui_copy_stub_clipboard = false
# If you are running gtl without X, the copy to clipboard feature
# will not work (Or if you don't have xclip or xsel).
# In this case, enabling this option will allow gtl to display
# the sub text in a modal for easy copy in tools like tmux
# At this stage, the rendering is ugly because of an issue in cview:
# https://code.rocketnine.space/tslocum/cview/issues/72#issuecomment-3968
tui_show_stub = false

## TUI Theming:
# Define colors for GTL TUI mode.
# Default (dark) theme:
#tui_color_background = "000000"
#tui_color_box = "FFFFFF"
#tui_color_focus_box = "008000"
#tui_color_author_name = "FF0000"
#tui_color_links = "87CEEB"
#tui_color_elapsed_time = "87CEEB"
#tui_color_text = "FFFFFF"
#tui_color_selected_background = "FFFFFF"
#tui_color_selected_foreground = "000000"
#tui_color_highlight = "FF0000"
#tui_color_quote = "808080"

# Bookmarks (only available via TUI for now):
bookmarks_enabled = false
bookmarks_file_path = "~/.config/gtl/bookmarks.json"
bookmarks_reverse_order = true
`)

func Init(configArg string, Data *core.TlData) {
	Config := getTlConfig(configArg)
	Data.Config = &Config

	if configureLogs(Config) != nil {
		log.Fatalln("Log file couldn't be created")
	}

	Feeds := getFeeds(Config.Subscribed_data)
	Data.Feeds = Feeds

	if Data.Config.Bookmarks_enabled {
		log.Println("Trying to enable bookmarks")
		filepath, err := loadBookmarkFile(Data.Config.Bookmarks_file_path)
		if err != nil {
			log.Println("Couldn't load bookmarks", err)
			log.Println("Disabling bookmarks")
			Data.Config.Bookmarks_enabled = false
		}

		bookmarks, e := loadBookmarks(filepath)
		if e != nil {
			log.Println("Couldn't init bookmarks", e)
			log.Println("Disabling bookmarks")
			Data.Config.Bookmarks_enabled = false
		}

		// Update real path to file in config object for reuse:
		Data.Config.Bookmarks_file_path = filepath
		// Save bookmarks in Data.
		Data.Bookmarks = &bookmarks
	}
}

// Get TlConfig.
func getTlConfig(configArg string) core.TlConfig {
	configFile, err := getConfigFilePath(configArg, "gtl.toml")
	if err != nil {
		log.Fatalln(err)
	}

	Config := core.TlConfig{}
	err = loadConfig(configFile, &Config)
	if err != nil {
		log.Fatalln(err)
	}

	return Config
}

// Load configuration file.
func loadConfig(configFile string, Config *core.TlConfig) error {
	file, err := os.Open(configFile)
	if err != nil {
		return fmt.Errorf("Couldn't open configuration file (%v)\n%v: ", configFile, err)
	}
	defer func() error {
		if err = file.Close(); err != nil {
			return fmt.Errorf("Couldn't close configuration file (%v)\n%v: ", configFile, err)
		}
		return nil
	}()

	b, err := ioutil.ReadAll(file)
	if err != nil {
		return fmt.Errorf("Couldn't read configuration file; ", configFile)
	}

	return toml.Unmarshal(b, Config)
}

// Init Config:
// - Try to load the given file if any.
// - Load default configuration file otherwise. Create it if it doesn't exist.
func getConfigFilePath(configArg string, defaultFilename string) (string, error) {
	var configFile string
	if configArg != "" {
		log.Println("Trying to load provided configuration file (%v)", configArg)
		if e := fileExist(configArg); e != nil {
			return "", e
		}

		configFile = configArg
		log.Println("Configuration file found:", configFile, "\n")

	} else {
		homepath, err := homedir.Dir()
		if err != nil {
			return "", fmt.Errorf("Error finding home directory")
		}

		configDir := filepath.Join(homepath, ".config", "gtl")
		if err = fileExist(configDir); err != nil {
			log.Println("GTL directory doesn't exist\n", err)
			os.Mkdir(configDir, 0744)
		}

		configFile = filepath.Join(configDir, defaultFilename)

		// Load or create configFile.
		log.Println("Trying to load default configuration file (%v)", configFile)
		f, err := os.OpenFile(configFile, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0666)
		defer f.Close()
		if err == nil {
			log.Println("Default configuration (" + defaultFilename + ") file does not exist yet, creating it…")
			_, err := f.Write(defaultConf)
			if err != nil {
				return "", fmt.Errorf("Default configuration file (" + defaultFilename + ") couldn't be created in default place")
			}
		}
	}

	return configFile, nil
}

// Check if a file exist and if not return a custom error.
func fileExist(filename string) error {
	_, err := os.Stat(filename)
	if err != nil {
		return fmt.Errorf("File (%v) not found", filename)
	}
	return nil
}

// Configure Log file.
func configureLogs(config core.TlConfig) error {
	var logFile string
	if config.Log_file != "" {
		logFile = config.Log_file
	} else {
		logFile = "gtl.log"
	}

	file, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
	if err != nil {
		return err
	}
	log.SetOutput(file)

	return nil
}

// Configure Bookmarks.
func loadBookmarks(filepath string) (core.TlBookmarks, error) {
	bookmarks, err := core.LoadBookmarksFromFile(filepath)
	return bookmarks, err
}

func loadBookmarkFile(bookmarkFile string) (string, error) {
	bookmarks_path, err := homedir.Expand(bookmarkFile)
	if err != nil {
		return "", err
	}

	filepath, err := getConfigFilePath(bookmarks_path, "bookmarks.json")

	// Given bookmark file that doesn't exist…
	// Let's create it!
	if len(bookmarkFile) > 0 && err != nil {
		bfp, e := homedir.Expand(bookmarkFile)
		if e != nil {
			return "", e
		}

		f, err := os.Create(bfp)
		defer f.Close()
		if err == nil {
			log.Println("Creating custom bookmarks file: " + bookmarkFile)
			filepath = bfp
		}
	}

	return filepath, nil
}
